designs/node-repair.md

```diff
@@ -0,0 +1,182 @@
+            TrolorationDuration: "10m"
+        },
+        ...
+    }
+}
+
+In the example above AWS Karpenter Provider supports monitoring and terminating two node status condition of the Kubelet Ready condition and the NPD  NetworkUnavailable condition. Below are the two cases of when we will act nodes:
+
+```
+apiVersion: v1
+kind: Node
+metadata:
+  ...
+status:
+  condation:
+    - lastHeartbeatTime: "2024-11-01T16:29:49Z"
+      lastTransitionTime: "2024-11-01T15:02:48Z"
+      message: no connection
+      reason: Network is not available
+      status: "False"
+      type: NetworkUnavailable
+    ...
+    
+- The Node here will be eligible for node repair after at `2024-11-01T15:12:48Z`    
+---
+apiVersion: v1
+kind: Node
+metadata:
+  ...
+status:
+  condation:
+    - lastHeartbeatTime: "2024-11-01T16:29:49Z"
+      lastTransitionTime: "2024-11-01T15:02:48Z"
+      message: kubelet is posting ready status
+      reason: KubeletReady
+      status: "False"
+      type: NetworkUnavailable
+    - lastHeartbeatTime: "2024-11-01T16:29:49Z"
+      lastTransitionTime: "2024-11-01T15:02:48Z"
+      message: kubelet is posting ready status
+      reason: KubeletReady
+      status: "False"
+      type: Ready
+    ...
+    
+- The Node here will be eligible for node repair after at `2024-11-01T15:32:48Z`    
+```
+
+
+### Option 2: Cloud provider interface for determining node health
+
+```
+type UnhealthyReason string
+
+type CloudProvider interface {
+  ...
+    // IsHealthy returns whether a NodeClaim is considered operating as expected
+    IsHealthy(context.Context, *v1.NodeClaim) (UnhealthyReason, error)
+  ...
+}
+```
+
+`IsHealthy` cloud provide methods allows each implementation of Karpenter to determine the healthy state of a NodeClaim and return a specified `UnhealthyReason` indicating that an action on the node is required. `IsHealthy` allows the cloud providers to determine when and on what condition to act on cleaning up the node. Cloud providers will also need to consider short lived failure prior to determining if a node is unhealthy. The `IsHealthy` follows the format defined by `IsDrifted`. 
+
+Main drawback of this option is providers will need to implement much more undifferentiated logic, without having the ability to model duration and remediation on nodes. The reconciliation step for replacing an unhealthy node will look like: 
+
+1. Karpenter will reconcile on the NodeClaim and make a cloud provide call with IsHealthy
+2. If the node is determined to unhealthy, a status condition will be added to the NodeClaim
+3. Disruption controller will force terminate the node claim immediately 
+
+```
+apiVersion: karpenter.sh/v1
+kind: NodeClaim
+metadata:
+  ...
+status:
+  conditions:
+    - lastTransitionTime: "2024-11-04T21:06:06Z"
+      status: "True"
+      reason: <cloud provider returned UnhealthyReason>
+      type: Unhealthy
+    ...
+```
+
+### Option 3: Cloud provider injected unhealthy condition 
```

Comment:
This also doesn't feel like a real option -- given that we are considering forceful termination here after a toleration duration, it feels like the real options are:
1. Each CloudProvider implements this behavior themselves and does the force termination on their own (we did this with the interruption controller)
2. The neutral code implements this behavior and the forceful termination is driven through the neutral code
3. There is some in-between option with `IsHealthy` and a graceful termination option where there is more value in pushing things into the neutral code.

There might also be more value talking about the upside/downside of why we are doing forceful termination and why we are proposing adding a toleration duration

Reply:
Dropping this option after an offline discussion